source/fundamentals/monitoring.txt

```diff
@@ -0,0 +1,115 @@
+subscribe to these driver events to monitor your application.
+
+.. note:: Event Logging
+
+   This page explains how to monitor your application in code. To learn how to record
+   this information to an external log, see :ref:`Logging .<csharp-logging>`
+
+.. _csharp-event-types:
+
+Event Types
+-----------
+
+The type of event that the driver emits depends on the operation being performed.
+The following table describes the types of events that the driver emits:
+
+.. list-table::
+
+   * - Event Type
+     - Description
+   * - Command Events
+     - Events related to MongoDB database commands, such as ``find``, ``insert``,
+       ``delete``, and ``count``. To learn how to use the {+driver-short+} to run a
+       database command, see :ref:`<csharp-run-command>`. For more information about
+       MongoDB database commands, see :manual:`Database Commands </reference/command/>`
+       in the {+mdb-server+} manual.
+
+       .. important:: Redacted Output
+
+          As a security measure, the driver redacts the contents of some command events. This
+          protects the sensitive information contained in these command events. For a
+          full list of redacted command events, see the
+          :spec:`MongoDB command logging and monitoring specification </command-logging-and-monitoring/command-logging-and-monitoring.rst#security>`.
+
+   * - Server Discovery and Monitoring (SDAM) Events
+     - Events related to changes in the state of the MongoDB instance or cluster.
+
+   * - Connection Pool Events
+     - Events related to the connection pool held by the driver.
+
+For a complete list of events the driver emits, see the API documentation for the
+`MongoDB.Driver.Core.Events <{+new-api-root+/MongoDB.Driver/MongoDB.Driver.Core.Events.html>`__
+namespace.
+
+.. _csharp-monitor-events:
+
+Subscribing to Events
+---------------------
+
+To monitor an event, you must subscribe a listener method on your ``MongoClient`` instance.
+The following steps describe how to subscribe to events:
+
+1. Create a ``MongoClientSettings`` object.
+#. Set the ``ClusterConfigurator`` property on the ``MongoClientSettings`` object to a
+   lambda function that accepts a ``ClusterBuilder`` object.
+#. In the lambda function, call the ``Subscribe<TEvent>()``
+   method on the ``ClusterBuilder`` object for each event you want to subscribe to.
+   Replace ``TEvent`` with the event type. Pass the event handler
+   method as an argument to the ``Subscribe<TEvent>()`` method.
+
+The following code example shows how to subscribe to the ``ClusterOpenedEvent``,
+``ServerHeartbeatSucceededEvent``, and ``ConnectionPoolReadyEvent``. This example
+assumes that the ``ClusterEventHandler``, ``HeartbeatEventHandler``,
+and ``ConnectionPoolEventHandler`` methods are defined elsewhere in your code.
+
+.. code-block:: csharp
+
+   var clientSettings = MongoClientSettings.FromConnectionString(MongoConnectionString);
```

Comment:
Tech reviewer: Does this process look any different in pre-3.0 driver versions? Or can I backport this all the way to 2.19?

Reply:
Users should have been using this event-subscription process even prior 3.0 driver. There is only one case where users could subscribe to a ClusterDescriptionChangedEvent directly through Client.Cluster since the cluster used to expose its DescriptionChanged handler. But we removed that in 3.0 and there are no other ways for users to subscribe to events other than through ClusterBuilder.

Reply:
Thanks. I'll backport this page all the way to v2.22 then.